Handle tensor default func args when inlining

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90575
* #91729

    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/pytorch/issues/93471    

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

wconstab committed:
Handle tensor default func args when inlining

Previously, when inlining a function, its default arguments
were only wrapped with VariableTrackers if non-tensor. Now,
tensor default args are also handled by adding them to the
parent InstructionTranslator as an attribute.

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


Previously, when inlining a function, its default arguments
were only wrapped with VariableTrackers if non-tensor. Now,
tensor default args are also handled by adding them to the
parent InstructionTranslator as an attribute.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


Previously, when inlining a function, its default arguments
were only wrapped with VariableTrackers if non-tensor. Now,
tensor default args are also handled by adding them to the
parent InstructionTranslator as an attribute.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


Previously, when inlining a function, its default arguments
were only wrapped with VariableTrackers if non-tensor. Now,
tensor default args are also handled by adding them to the
parent InstructionTranslator as an attribute.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/output_graph.py:
```diff
@@ -560,7 +560,7 @@ def compile_and_call_fx_graph(self, tx, rv, root):
                 )
                 log.log(
-                    logging.CODE,  # type: ignore[attr-defined]
+                    logging.INFO,  # type: ignore[attr-defined]
```
separate PR plz

wconstab replied:
oh its already landing
https://github.com/pytorch/pytorch/pull/90572

voznesenskym commented on torch/_dynamo/variables/functions.py:
```diff
@@ -30,16 +33,38 @@ def wrap_bound_arg(val, options):
         return variables.EnumVariable(val, **options)
     elif isinstance(val, (type, abc.ABCMeta)):
         return variables.UserDefinedClassVariable(val, **options)
+    elif istensor(val):
```
I hate this util

wconstab replied:
why, is it dangerous? should i be specific about which tensor classes i support here?

voznesenskym commented on torch/_dynamo/variables/functions.py:
```diff
@@ -30,16 +33,38 @@ def wrap_bound_arg(val, options):
+        # Assumption: this is only called from InliningInstructionTranslator, in which case
+        # it is assured we will not 'reconstruct' this value in codegen, since the usage is
+        # entirely inlined.
+        # TODO - can we assert this? we expect to not reach this case for non-inline,
+        # since InstructionTranslator __init__ prepares VariableTrackers for args of top
+        # level function including defaults.
```
Yes you can - check `_produce_guard_code` in guards has a prior example of assuring a caller.

wconstab replied:
oh, crawling stack frames lol. yes.

otoh, i wrote that comment before i figured out how to set up the source/guard correctly.  now, i'm not sure if it is important to assert this or state the assumption.  i am thinking this codepath should work generally now?

wconstab replied:
no longer an issue

voznesenskym commented on torch/_dynamo/variables/functions.py:
```diff
@@ -30,16 +33,38 @@ def wrap_bound_arg(val, options):
+        # a NNModuleVariable, currently the UserFunctionVariable itself has no source
+        # so we need to fix that first
+        # assert source is not None, "source must be provided for tensor arg"
+        if source:
+            return tx.output.register_attr_or_module(
+                val,
+                # should there be a "name" here?
+                source=source,
+            )
+        else:
+            return tx.output.register_attr_or_module(
+                val,
+            )
```
Sorry, when is source None here?

wconstab replied:
ok, this part should be resolved in the latest code.  I added a source to nnmodule call function, so that i could then build an appropriate source here.

voznesenskym commented on torch/_dynamo/variables/functions.py:
```diff
@@ -30,16 +33,38 @@ def wrap_bound_arg(val, options):
 
 
-def wrap_args_kwargs(result, options):
+def wrap_args_kwargs(result, options, tx):
```
nit: tx usually goes first, unless its onerous, just do that

voznesenskym commented on torch/_dynamo/variables/functions.py:
```diff
@@ -117,28 +142,44 @@ def get_globals(self):
 
         fn: types.FunctionType = self.fn
+        # TODO comment
+        defaults = fn.__defaults__ or []
+        defaults_source = AttrSource(self.source, "__defaults__")
+        defaults_item_sources = [
+            GetItemSource(defaults_source, i) for i, _ in enumerate(defaults)
+        ]
```
You do this basically twice, once for defaults and once for kw, consider if there's a helper fn somewhere here, esp so you dont have to zip in the list comprehension below.

wconstab replied:
Yea, i actually coded up a helper but it was not convenient since i wanted to output a list for defaults and a dict for kwdefaults.  Now that I switched to a specialized DefaultsSource, it encapsulates some of the stuff and this code seems good to me without an additional helper.

voznesenskym commented on torch/_dynamo/variables/functions.py:
```diff
@@ -30,16 +33,38 @@ def wrap_bound_arg(val, options):
+        # so we need to fix that first
+        # assert source is not None, "source must be provided for tensor arg"
+        if source:
+            return tx.output.register_attr_or_module(
```
`VariableBuilder(tx, source)(val)` would be a better entry point, would that work for you?

wconstab replied:
yes, it seems to work fine.  i thought it did not work but i was probably holding it wrong before.

there is, annoyingly, a circular import

wconstab committed:
Update on "Handle tensor default func args when inlining"


Previously, when inlining a function, its default arguments
were only wrapped with VariableTrackers if non-tensor. Now,
tensor default args are also handled by adding them to the
parent InstructionTranslator as an attribute.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


Previously, when inlining a function, its default arguments
were only wrapped with VariableTrackers if non-tensor. Now,
tensor default args are also handled by adding them to the
parent InstructionTranslator as an attribute.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


Previously, when inlining a function, its default arguments
were only wrapped with VariableTrackers if non-tensor. Now,
tensor default args are also handled by adding them to the
parent InstructionTranslator as an attribute.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function
    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function
    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented:
ok @voznesenskym i think its ready for a review. still outstanding is what to do with the following

                and "__defaults__" not in guard.name
                and "__kwdefaults__" not in guard.name

Although if the solution isn't fairly straightforward, maybe we can land as-is

voznesenskym commented on torch/_dynamo/source.py:
```diff
@@ -148,6 +148,45 @@ def name(self):
+        super().__init__()
+        self.base = base
+        self.idx_key = idx_key
+        self.is_kw = is_kw
+        self.field = "__kwdefaults__" if is_kw else "__defaults__"
+        if not is_kw:
+            assert isinstance(idx_key, int)
+        else:
+            assert isinstance(idx_key, str)
```
nit:
```suggestion
        self.is_kw = is_kw
        if self.is_kw:
            assert isinstance(idx_key, str)
            self.field = "__kwdefaults__"
        else:
            assert isinstance(idx_key, int)
            self.field = "__defaults__"
```

voznesenskym commented on torch/_dynamo/source.py:
```diff
@@ -148,6 +148,45 @@ def name(self):
+        return self.base.guard_source()
+
+    def name(self):
+        if self.is_kw:
+            return f"{self.base.name()}.__kwdefaults__['{self.idx_key}']"
+        else:
+            return f"{self.base.name()}.__defaults__[{self.idx_key}]"
```
nit: pull it up to the ctor, and just make this return something like `self.name?`

voznesenskym commented on torch/_dynamo/variables/functions.py:
```diff
@@ -30,16 +30,24 @@ def wrap_bound_arg(val, options):
     elif isinstance(val, (type, abc.ABCMeta)):
         return variables.UserDefinedClassVariable(val, **options)
+    elif istensor(val):
+        # Circular import...
```
rm comment, we do this all over anyway

voznesenskym commented on torch/_dynamo/variables/functions.py:
```diff
@@ -30,16 +30,24 @@ def wrap_bound_arg(val, options):
+        assert (
+            source is not None
+        ), "Must provide a source if wrapping a tensor arg, otherwise can't guard"
+        return VariableBuilder(tx, source)(val)
```
glad this worked out

voznesenskym commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -212,11 +212,14 @@ def record_nn_module_stack():
                 # for lazy modules, run the pre-hooks which will update the type
                 # TODO mlazos: we don't fully support all of the hooks that exist,
                 # so restrict using __call__ only to lazy modules for now
+                class_source = AttrSource(self.source, "__class__")
```
should we assert source here?

wconstab replied:
can do. not sure which assert you meant but i added 2.  probably the self.source one is the more critical but the second one can't hurt

voznesenskym commented on torch/_dynamo/variables/functions.py:
```diff
@@ -369,6 +389,7 @@ def get_globals(self):
 
     def bind_args(self, parent, args, kwargs):
         code = self.get_code()
+        tx = parent.output.root_tx
```
made on L392, used on L406, move it down? Or even just slap it right into the only line that uses it?

wconstab replied:
yea i can simplify. did this on purpose to copy the pattern for the other 'bind_args'

voznesenskym reviewed (approved):
Bunch of nits, LGTM, maybe nag @jansel for a peek too?

voznesenskym commented on torch/_dynamo/guards.py:
```diff
@@ -455,7 +455,15 @@ def combine_scopes(left, right):
         global_builder = GuardBuilder(self.id_ref, f_globals, self, renames=False)
         for guard in sorted(guards or [], key=Guard.sort_key):
-            if not config.guard_nn_modules and guard.is_nn_module():
+            if (
+                not config.guard_nn_modules
+                and guard.is_nn_module()
+                # we must guard on default args for functions,
+                # even if we specialize on the rest of the nnmodule contents
+                # TODO(whc) - presumably we can use the 'DefaultsSource' in some way here instead of name parsing?
+                and "__defaults__" not in guard.name
+                and "__kwdefaults__" not in guard.name
+            ):
```
This is not actually as offensive as I thought it would be. We could use tagging here, or a new kind of guard (then you can check those as types).

One thing I'm wondering - if these are `DefaultsSource` now, we could always detect that when we make the guard, and write something like `guard.is_default()` as an analogue to `guard.is_nn_module()` (Or even add a guardsource subtype?). 

No strong opinion, but if this read:

```
if (
                not config.guard_nn_modules
                and guard.is_nn_module()
                and guard.is_default()
```

That would be way cleaner :)

wconstab replied:
hmm, you can't extend an enum (GuardSource) right? and i'm not sure if it makes sense to add a new enum value for defaults, since defaults can also come from a global function or a local function etc. but if it does make sense, i could add GuardSource.DEFAULTS and go that way. i'll keep this as a next PR and land the rest as is.

voznesenskym replied:
Sorry, extend as in add a new value to, not via inheritance, but doing it in a later PR is fine too.

voznesenskym replied:
If we leave it as is, let's add a comment referencing DefaultSource.

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/source.py:
```diff
@@ -148,6 +148,46 @@ def name(self):
+
+    def __init__(self, base, idx_key, is_kw=False):
+        super().__init__()
+        assert base, "Must provide a valid base source"
```
hyper-nit Let's add a longer comment here, I could see folks hitting this and not understanding what a source is?

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

mlazos commented:
I don't understand why we need to guard on default args at all .. shouldn't we instantiate them on each call? It's equivalent to having 
```
f(x=None):
  if x is None:
    x = <default value>()
```
The variable tracker for default value shouldn't have a source and that's okay.

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Handle tensor default func args when inlining"


    Handle tensor default func/method args when inlining
    
    Previously, when inlining a function, its default arguments
    were only wrapped with VariableTrackers if non-tensor. Now,
    tensor default args are also handled by adding them to the
    parent InstructionTranslator as an attribute.
    
    - also patches up a missing source in nnmodule call_function,
      needed to properly guard on a default arg in its methods
    - adds new 'DefaultsSource' type which guards either a `__defaults__`
      or `__kwdefaults__` entry on a function

Fixes #90361  https://github.com/pytorch/torchdynamo/issues/1968    

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge -f "flaky ci only"

ezyang commented on torch/_dynamo/variables/functions.py:
```diff
@@ -9,37 +9,58 @@
+    elif istensor(val):
+        from torch._dynamo.variables.builder import VariableBuilder
+
+        return VariableBuilder(tx, source=source, **options)(val)
```
This doesn't make much sense to me, VariableBuilder doesn't actually take any arguments besides tx and source...
